File: app/src/main/java/com/example/myapplication/leetcode10.java
package com.example.myapplication;

public class leetcode10 {

    public boolean isMatch(String s, String p) {
        char[] sChars = s.toCharArray();
        char[] pChars = p.toCharArray();

        // p匹配的范围可能是大于s的，所以要从 p 开始遍历，当第一个位置符合时候，进入匹配，如果最后都 index 都还没有==s的index ，那么就匹配失败了，否则成功

        int i = 0;

        return true;
    }

}

File: app/src/main/java/com/example/myapplication/leetcode10.kt
package com.example.myapplication

class leetcode10 {
    fun isMatch(s: String, p: String): Boolean {
        var isAll = false
        var sCursor = 0
        var pCursor = 0
        for (c in s.indices) {

            try {
                val sChar = s[sCursor]
                val pChar = p[pCursor]

                if (sChar == pChar) {
                    //通过检测，检测下一位
                    sCursor++
                    pCursor++
                    continue
                }

                if (pChar == '.') {
                    //全匹配
                    sCursor++
                    pCursor++
                    continue
                }

                if (pChar == '*') {
                    if (s[sCursor] == p[pCursor - 1] || p[pCursor - 1] == '.') {
                        isAll = true
                        sCursor++
                        continue
                    } else {
                        if(isAll){
                            isAll = false
                            pCursor++
                            continue
                        }else{
                            return false
                        }
                    }
                }
                return false
            } catch (e: Exception) {
                return false
            }
        }
        return true
    }
}